client/Lykke.AlgoStore.MatchingEngineAdapter.Client/RequestManager.cs

```diff
@@ -0,0 +1,84 @@
+﻿using Lykke.AlgoStore.MatchingEngineAdapter.Abstractions.Domain.Listening.Requests;
+using Lykke.AlgoStore.MatchingEngineAdapter.Abstractions.Services.Listening;
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Lykke.AlgoStore.MatchingEngineAdapter.Client
+{
+    internal class RequestManager : IRequestManager
+    {
+        private readonly IMeaCommunicator _meaCommunicator;
+        private readonly ConcurrentDictionary<uint, TaskCompletionSource<IMessageInfo>> 
+            _queuedMessages = new ConcurrentDictionary<uint, TaskCompletionSource<IMessageInfo>>();
+
+        private string _clientId;
+        private string _instanceId;
+
+        private int _currentRequestId = -1;
+
+        public RequestManager(IMeaCommunicator meaCommunicator)
+        {
+            _meaCommunicator = meaCommunicator ?? throw new ArgumentNullException(nameof(meaCommunicator));
+            _meaCommunicator.OnMessageReceived += ProcessResponse;
+            _meaCommunicator.OnConnectionEstablished += SendAuthRequest;
+        }
+
+        public void SetClientAndInstanceId(string clientId, string instanceId)
+        {
+            _clientId = clientId;
+            _instanceId = instanceId;
+
+            _meaCommunicator.Start();
+        }
+
+        public Task<IMessageInfo> MakeRequestAsync<T>(MeaRequestType requestType, T message)
+        {
+            var taskCompletionSource = new TaskCompletionSource<IMessageInfo>();
+            var nextRequestId = GetNextRequestId();
+
+            _queuedMessages.TryAdd(nextRequestId, taskCompletionSource);
+            // Wait here because otherwise we can't return the task from the completion source
+            _meaCommunicator.SendRequestAsync(nextRequestId, (byte)requestType, message).Wait();
```

Comment:
await _meaCommunicator.SendRequestAsync(nextRequestId, (byte)requestType, message);

you should not use Wait() in async methods